### DISCUSSION TITLE
obs cannot capture screen

### DISCUSSION THREAD

**Discussion by tomysurya** [YaLTeR/niri#309] [Q&A] [answered]
hey, i love this compositor, i am using it daily, but obs capture screen not working, already install all the requirements xdg-desktop-portal-gnome, pipewire, set env etc. but option to capture screen pipewire is missing. what did i miss?

**Comment by YaLTeR** [comment]
Hey, can you attach niri output? You need to start niri as a session (via systemd, or `niri --session`), make sure pipewire is running beforehand, and make sure xdp-gnome is running afterwards.

**Reply by tomysurya** [reply]
okay i'll try but which package is xdp-gnome? is it came with xdg-dekstop-portal-gnome? i'm on arch linux

**Reply by YaLTeR** [reply]
that's just a short way of writing xdg-dekstop-portal-gnome

**Reply by tomysurya** [reply]
![ ss_12:03--Sat-27-2024](https://github.com/YaLTeR/niri/assets/152084609/8fcfc614-ed13-4793-bea7-2566b221945c)
still not working, pipewire, xdg-desktop-portal-gnome already running, installed all the resources files to right place,  and i'm using tty to start niri --session

**Reply by YaLTeR** [reply]
You should have "Screen Capture (PipeWire)" in OBS. I'm not sure it's niri-related whether it's there or not, could you be missing some part of OBS? I'm using OBS from Flathub FWIW

**Reply by tomysurya** [reply]
yes thats what i mean, but screen capture (pipewire) in hyprland it does appear and worked

**Reply by YaLTeR** [reply]
Maybe OBS has something in the output?

**Comment by tomysurya** [comment]
i installed gnome-shell, and then it works, but when gnome-shell deleted, it's not work

**Reply by YaLTeR** [reply]
huh, guess you're missing something then

**Reply by tomysurya** [reply, marked as answer]
finally it worked! without gnome-shell, what i do is install xdg-desktop-wlr then start niri from tty with niri-session not niri --session, finally this compositor is complete for me, i use it as a daily driver, ditching hyprland for the scrolling tiles!!

**Reply by YaLTeR** [reply]
Great to hear that it worked! Though, niri does not use xdg-desktop-portal-wlr, so it is likely some other component that you were missing.

**Comment by rrveex** [comment]
This was the important part for me (no systemd):

> ... make sure pipewire is running beforehand ...

**Don't** `spawn-at-startup "pipewire"` (this works in hyprland, not in niri); instead:

```
dbus-run-session pipewire &
# sleep 0.5  # try this if it doesn't work without
dbus-run-session niri --session
```

`portal-gnome` seems to be automatically started by `niri`.

I believe the only purpose of systemd-stuff in niri is just to "start pipewire *before* portal-gnome". Sad to see gentoo/guru doing this:

```
IUSE="+dbus screencast systemd"
REQUIRED_USE="
	screencast? ( dbus )
	systemd? ( dbus )
"
```

**No**, actually it works perfectly fine without.

**Comment by janhohenheim** [comment]
I had the same issue, and the following worked for me:
```bash
systemctl --user add-wants niri.service xdg-desktop-portal-gtk.service
systemctl --user add-wants niri.service xdg-desktop-portal-gnome.service
systemctl --user add-wants niri.service pipewire.service
```

**Comment by tomsch** [comment]
## Fix: OBS "Screen Capture (PipeWire)" on Niri

  **Environment:** NixOS 25.05, Niri (niri-flake), OBS from nixpkgs (not Flatpak)

  ### Required Services

  - `xdg-desktop-portal`
  - `xdg-desktop-portal-wlr` (ScreenCast, Screenshot)
  - `xdg-desktop-portal-gtk` (FileChooser, Settings, OpenURI)
  - `pipewire` + `wireplumber`

  Note: `xdg-desktop-portal-gnome` is **not needed** and doesn't work on Niri anyway.

  ### Root Cause

  1. `wlr.portal` has `UseIn=wlroots;sway;Wayfire;river;phosh;Hyprland;` — **no `niri`**
  2. User portal config `~/.config/xdg-desktop-portal/portals.conf` didn't specify `ScreenCast=wlr`

  ### Fix

  **1. NixOS config** — Set `XDG_CURRENT_DESKTOP=sway` for wlr portal:

  ```nix
  systemd.user.services.xdg-desktop-portal-wlr.environment = {
    WAYLAND_DISPLAY = "wayland-1";
    XDG_CURRENT_DESKTOP = "sway";
  };

  2. User config — Create ~/.config/xdg-desktop-portal/portals.conf:

  [preferred]
  default=gtk
  org.freedesktop.impl.portal.ScreenCast=wlr
  org.freedesktop.impl.portal.Screenshot=wlr

  3. Restart:

  systemctl --user restart xdg-desktop-portal xdg-desktop-portal-wlr

  Verify

  busctl --user introspect org.freedesktop.portal.Desktop /org/freedesktop/portal/desktop | grep ScreenCast

**Reply by Sighthesia** [reply]
Thank you! Now I finally understand that this is because OBS is using the wrong xdg-desktop-portal.

**Reply by SeniorMatt** [reply]
> ## Fix: OBS "Screen Capture (PipeWire)" on Niri
> 
> **Environment:** NixOS 25.05, Niri (niri-flake), OBS from nixpkgs (not Flatpak)
> ### Required Services
> 
>     * `xdg-desktop-portal`
> 
>     * `xdg-desktop-portal-wlr` (ScreenCast, Screenshot)
> 
>     * `xdg-desktop-portal-gtk` (FileChooser, Settings, OpenURI)
> 
>     * `pipewire` + `wireplumber`
> 
> 
> Note: `xdg-desktop-portal-gnome` is **not needed** and doesn't work on Niri anyway.
> ### Root Cause
> 
>     1. `wlr.portal` has `UseIn=wlroots;sway;Wayfire;river;phosh;Hyprland;` — **no `niri`**
> 
>     2. User portal config `~/.config/xdg-desktop-portal/portals.conf` didn't specify `ScreenCast=wlr`
> 
> 
> ### Fix
> 
> **1. NixOS config** — Set `XDG_CURRENT_DESKTOP=sway` for wlr portal:
> 
> ```nix
> systemd.user.services.xdg-desktop-portal-wlr.environment = {
>   WAYLAND_DISPLAY = "wayland-1";
>   XDG_CURRENT_DESKTOP = "sway";
> };
> 
> 2. User config — Create ~/.config/xdg-desktop-portal/portals.conf:
> 
> [preferred]
> default=gtk
> org.freedesktop.impl.portal.ScreenCast=wlr
> org.freedesktop.impl.portal.Screenshot=wlr
> 
> 3. Restart:
> 
> systemctl --user restart xdg-desktop-portal xdg-desktop-portal-wlr
> 
> Verify
> 
> busctl --user introspect org.freedesktop.portal.Desktop /org/freedesktop/portal/desktop | grep ScreenCast
> ```

It doesn't work for me on NixOS unstable for some reason. 

> busctl --user introspect org.freedesktop.portal.Desktop /org/freedesktop/portal/desktop | grep ScreenCast

it shows nothing

**Reply by prothegee** [reply]
```
2. User config — Create ~/.config/xdg-desktop-portal/portals.conf:

[preferred]
default=gtk
org.freedesktop.impl.portal.ScreenCast=wlr
org.freedesktop.impl.portal.Screenshot=wlr
```

this one work in arch, just make it sure that those desktop deps and related pipewire  installed

**Reply by YaLTeR** [reply]
I suggest *not* using the wlr portal for screencast, it is really quite less optimal than what you get with the gnome portal, at least currently

**Comment by tomsch** [comment]
i switched back to gnome portal and now all is working (also discord screen sharing)

```
extraPortals = [
    pkgs.xdg-desktop-portal-gtk
    pkgs.xdg-desktop-portal-gnome
  ];
  ScreenCast = [ "gnome" ];
  XDG_CURRENT_DESKTOP = "niri";
```

**Comment by aksiksi** [comment]
In my case, I had to include `GNOME` as part of `XDG_CURRENT_DESKTOP` explicitly:

```kdl
environment {
    // "Trick" OBS into believing that this a Gnome desktop for screen recording
    // to work.
    XDG_CURRENT_DESKTOP "niri:GNOME"
}
```

With this, the Pipewire screen recording option appears in OBS.

**Comment by Rosnerv** [comment]
Hi,

I have the same problem where OBS is not capturing the screen, but I’m on KaOS and the package xdg-desktop-portal-wlr is not available here.

What alternative do I have?

**Reply by prothegee** [reply]
use desktop portal gnome,
try search install something like `xdg-desktop-portal xdg-desktop-portal-gnome xdg-desktop-portal-gtk` in your package disto

then check those user systemctl
```sh
systemctl --user status xdg-desktop-portal
systemctl --user status xdg-desktop-portal-gtk
systemctl --user status xdg-desktop-portal-gnome
```

also try add this file in `~/.config/xdg-desktop-portal`
```
[preferred]
default=gtk;
org.freedesktop.impl.portal.ScreenCast=gnome;
# org.freedesktop.impl.portal.Screenshot=wlr;
```

then daemon-reload and restart those user service again.

**Comment by thiagovianav** [comment]
if someone stuck in this for some reason and gets a black screen when trying (this happened when I ran `yay -Syu`.
you can just run `yay -S xdg-desktop-portal-wlr` and you should be fine (as me)
